compiler/rustc_lint/src/builtin.rs

```diff
@@ -2420,3 +2420,37 @@ impl<'tcx> LateLintPass<'tcx> for ClashingExternDeclarations {
         }
     }
 }
+
+declare_lint! {
+    FUNCTION_REFERENCES,
+    Warn,
+    "suggest casting functions to pointers when attempting to take references"
+}
+
+declare_lint_pass!(FunctionReferences => [FUNCTION_REFERENCES]);
+
+impl<'tcx> LateLintPass<'tcx> for FunctionReferences {
+    fn check_expr(&mut self, cx: &LateContext<'_>, e: &hir::Expr<'_>) {
+        if let hir::ExprKind::AddrOf(hir::BorrowKind::Ref, _, referent) = e.kind {
+            if let hir::ExprKind::Path(qpath) = &referent.kind {
+                if let Some(def_id) = cx.qpath_res(qpath, referent.hir_id).opt_def_id() {
+                    cx.tcx.hir().get_if_local(def_id).map(|node| {
+                        node.fn_decl().map(|decl| {
+                            if let Some(ident) = node.ident() {
+                                cx.struct_span_lint(FUNCTION_REFERENCES, referent.span, |lint| {
+                                    let num_args = decl.inputs.len();
+                                    lint.build(&format!(
+                                        "cast `{}` with `as *const fn({}) -> _` to use it as a pointer",
```

Comment:
```suggestion
                                        "cast `{}` with `as fn({}) -> _` to use it as a pointer",
```

We definitely shouldn't suggest a pointer to a function pointer 😉.

It might also be worth including things like ABI, unsafety and whether it's a variadic function so the suggestion could be applicable in more cases.